torch_xla/experimental/tpu.py

```diff
@@ -0,0 +1,147 @@
+import functools
+import operator
+import os
+from typing import Dict, NamedTuple, Optional, List, Tuple
+import requests
+import yaml
+
+import torch_xla.utils.utils as xu
+import torch_xla.core.xla_env_vars as xenv
+
+_GCE_METADATA_ROOT_URL = 'http://metadata.google.internal/computeMetadata/v1'
+_ACCELERATOR_TYPE_TO_HOST_BOUNDS = {
+    # v2
+    'v2-8': '1,1,1',
+    'v2-32': '2,2,1',
+    'v2-128': '4,4,1',
+    'v2-256': '4,8,1',
+    'v2-512': '8,8,1',
+    # v3
+    'v3-8': '1,1,1',
+    'v3-32': '2,2,1',
+    'v3-64': '2,4,1',
+    'v3-128': '4,4,1',
+    'v3-256': '4,8,1',
+    'v3-512': '8,8,1',
+    'v3-1024': '8,16,1',
+    'v3-2048': '16,16,1',
+}
+
+
+class MeshShape(NamedTuple):
+  """Represents a TPU mesh shape (e.g. '2,2,1' or '1,1,1')"""
+  x: int
+  y: int
+  z: int
+
+  @classmethod
+  def from_string(cls, mesh: str):
+    dims = tuple(int(d) for d in mesh.split(','))
+    if len(dims) != 3:
+      raise ValueError("Mesh shape '{}' should be length 3".format(mesh))
+
+    return MeshShape(*dims)
+
+  @property
+  def size(self) -> int:
+    return functools.reduce(operator.mul, self)
+
+  def __mul__(self, other):
+    return MeshShape(*(d1 * d2 for d1, d2 in zip(self, other)))
+
+
+def _get_metadata(key: str) -> str:
+  path = os.path.join(_GCE_METADATA_ROOT_URL, 'instance/attributes', key)
+  resp = requests.get(path, headers={'Metadata-Flavor': 'Google'})
+  resp.raise_for_status()
+
+  return resp.text
+
+
+def process_bounds_size(default: int = 1) -> int:
+  """Returns number of processes across all TPU hosts."""
+  process_bounds = xu.getenv_as(xenv.TPU_PROCESS_BOUNDS, str)
+
+  return MeshShape.from_string(
+      process_bounds).size if process_bounds else default
+
+
+def num_local_processes(local_chips: int = 4) -> int:
+  """Returns number of processes to create on this host."""
+  # Don't create more processes than local chips
+  return min(local_chips, process_bounds_size(default=local_chips))
```

Comment:
Process bounds includes processes across all hosts. For example, the bounds for a v4-16 would be `2,2,2` (world size 8), but we can't create more than 4 processes on each host.

My logic here is:

- If process bounds are less than the number of local chips (`1,1,1,`), then only create that number of processes locally
- If the process bounds are larger than the number of local chips, only spawn at most one process per chip
- If the process bounds are unset, default to one process per chip

This falls through in the case where the user has some unusual configuration (say, two chips per process), but in that case, they should set the TPU environment variables themselves to explicitly assign visible chips to processes rather than have us try to infer their topology. We have a bug filed internally to add a `num_processes` flag to `pjrt.run_multiprocess` if a user needs to explicitly select some other number of local processes.